lib/elixir/lib/integer.ex

```diff
@@ -238,67 +238,51 @@ defmodule Integer do
   def parse("", base) when base in 2..36,
     do: :error
 
-  def parse(binary, base) when is_binary(binary) and base in 2..36 do
-    parse_in_base(binary, base)
+  def parse(<<bin::binary>>, base) when base in 2..36 do
+    parse_sign(bin, base)
```

Comment:
I feared so. Sad! It makes the code a bit convolute and it's really hard to get why just by reading it. Maybe we can add a comment about why it's done this way, not sure. We could in the meantime rename `parse_sign/2` to `parse_sign_and_digits/2`, wdyt?